src/Components/Aspire.Confluent.Kafka/ConfluentKafkaMetrics.cs

```diff
@@ -0,0 +1,90 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Concurrent;
+using System.Diagnostics.Metrics;
+
+namespace Aspire.Confluent.Kafka;
+
+internal sealed class ConfluentKafkaMetrics
+{
+    private readonly IMeterFactory _meterFactory;
+    private readonly Meter _meter;
+    public Counter<long> Tx { get; }
+    public Counter<long> TxBytes { get; }
+    public Counter<long> TxMessages { get; }
+    public Counter<long> TxMessageBytes { get; }
+    public Counter<long> Rx { get; }
+    public Counter<long> RxBytes { get; }
+    public Counter<long> RxMessages { get; }
+    public Counter<long> RxMessageBytes { get; }
+
+    public ConcurrentQueue<Measurement<long>> ReplyQueueMeasurements { get; } = new ConcurrentQueue<Measurement<long>>();
+    public ConcurrentQueue<Measurement<long>> MessageCountMeasurements { get; } = new ConcurrentQueue<Measurement<long>>();
+    public ConcurrentQueue<Measurement<long>> MessageSizeMeasurements { get; } = new ConcurrentQueue<Measurement<long>>();
+
+    public ConfluentKafkaMetrics(IMeterFactory meterFactory)
+    {
+        _meterFactory = meterFactory;
+        _meter = _meterFactory.Create(ConfluentKafkaCommon.MeterName);
+
+        _meter.CreateObservableGauge(Counters.ReplyQueue, GetReplyQMeasurements);
+        _meter.CreateObservableGauge(Counters.MessageCount, GetMessageCountMeasurements);
+        _meter.CreateObservableGauge(Counters.MessageSize, GetMessageSizeMeasurements);
+
+        Tx = _meter.CreateCounter<long>(Counters.Tx);
+        TxBytes = _meter.CreateCounter<long>(Counters.TxBytes);
+        TxMessages = _meter.CreateCounter<long>(Counters.TxMessages);
+        TxMessageBytes = _meter.CreateCounter<long>(Counters.TxMessageBytes);
+        Rx = _meter.CreateCounter<long>(Counters.Rx);
+        RxBytes = _meter.CreateCounter<long>(Counters.RxBytes);
+        RxMessages = _meter.CreateCounter<long>(Counters.RxMessages);
+        RxMessageBytes = _meter.CreateCounter<long>(Counters.RxMessageBytes);
+    }
+
+    public static class Counters
+    {
+        public const string ReplyQueue = "replyq";
```

Comment:
I'm not satisfied either by these names (basically same names as the ones from librdkafka [statistics.json](https://github.com/confluentinc/librdkafka/blob/master/STATISTICS.md)), but the naming pattern you linked seems to concern spans not metrics.
As these constants are used to expose specific metrics related to a given producer or consumer instance, I think we should instead follow a metric naming pattern.

Reply:
@lmolkova you are needed!

Reply:
for the time being I prefixed metrics with `messaging.kafka`

Reply:
Kafka metrics are in a weird place.
OTel is in the process of adding general messaging metrics into the spec - https://github.com/open-telemetry/semantic-conventions/pull/163 (will be merged soon).

However Kafka as a project decided (AFAIK) to report a set of client metrics ([KIP-714](https://cwiki.apache.org/confluence/display/KAFKA/KIP-714%3A+Client+metrics+and+observability)) that would be sent by default back to the broker and users would consume them from it. They decided not to follow otel semconv for the time being because of some back-compat reasons and limitations of existing metric impl on the broker side. Since they don't follow otel conventions, they also don't prefix with `messaging.` namespace.

I have a couple of suggestion 
- let's get an opinion from Kafka folks to understand the bigger picture.  @AndrewJSchofield what are your thoughts on Kafka integration component for .NET Aspire emitting it's own set of metrics? Also, could you confirm if I remember the decision (above) correctly or maybe something has changed? I envision the same conversation in OTel instrumentation world in Java and Python.

- in the short term, WRT metrics added here, I'd recommend aligning them with messaging semantic conventions when applicable and following otel naming conventions otherwise. I'll go through the changes and add my suggestions.

/cc @pyohannes

Reply:
I think we should align with the semantic conventions as well, even if they are not final. Maybe we can help push towards that goal.

Reply:
> I think we should align with the semantic conventions as well, even if they are not final. Maybe we can help push towards that goal.

we can't because we just don't get the right data from the kafka client library. 

We're going to discuss naming/approach with otel semconv community - https://github.com/open-telemetry/semantic-conventions/issues/578, I should have an update soon.

Reply:
@g7ed6e would you be able to take a quick look at https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/bbfe950ad0ace8123a5e6817fb3767e27a1a2cee/instrumentation/kafka/kafka-clients/kafka-clients-2.6/library/README.md and check if any of those metrics apply here and if we can unify the naming? (just discovered they exist).

In any case, metric naming and that discussion should not block this PR and can be changed in the follow-ups.

Reply:
@pyohannes : I updated and renamed `messaging.kafka.message.transmitted` to [messaging.publish.messages](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/messaging/messaging-metrics.md#metric-messagingpublishmessages) and 
`messaging.kafka.message.received` to [messaging.receive.messages](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/messaging/messaging-metrics.md#metric-messagingreceivemessages).

@lmolkova : I had a look to the linked implementation. Currently (in this Aspire PR) we do not intercept messages between user code and kafka client library. Instead we use a librdkafka feature which perdiodically fetch statistics (please notice we only parse the [top level properties from the statistics json](https://github.com/confluentinc/librdkafka/blob/master/STATISTICS.md) for now, there should be other metrics we may probably use and map to metrics in the future).
Regarding the interceptor / decorator approach we may implement it, but before I would like to ensure that we want to do this in Aspire component as having a look to the other components they do not hold too much code/logic and only exposes existing clients. 

@eerhardt / @davidfowl  : What are your thoughts about implementing otel distributed tracing with an internal wrapper around `IProducer<TKey, TValue>` / `IConsumer<TKey, TValue>` living in Aspire.Confluent.Kafka.

Reply:
> @eerhardt / @davidfowl : What are your thoughts about implementing otel distributed tracing with an internal wrapper around IProducer<TKey, TValue> / IConsumer<TKey, TValue> living in Aspire.Confluent.Kafka.

Yes, this sounds reasonable as a way to drive the conversation forward for the default .NET Kafka library.

@eerhardt I know we don't do this today with rabbit MQ (I think it's harder there because there are more APIs to wrap).